Web Zip file importing

BryonLewis opened the pull request:
fixes #1033 

Initial version of minimum viable product with basic functionality for uploading a zip containing flat list of media files and then running a post process on them to create a working annotation dataset.

Completed:

- Added in Zip upload Button to the upload menu.  This included changing the import-button component to have a small tag so the buttons better fit in the upload dialog.  A Secondary fact was making it so when uploading a zip file you can only specify the folder name and the annotation FPS.
- Post Process will now first look for a zip file which doesn't have the `ZipFileExtractedMarker` marker on it and will kick off a task to extract the zip file.  On subsequent calls it will ignore the zip files which have this marker preventing zip files from being extracted multiple times.
- Checks are in place for ZipBombs, based on the compression level of the zip file as well as nested Zip files.  This should hopefully prevent nefarious uploads of zip files from being used.
- Zip extraction for a single flat zip file works by checking for bad zip files (Nested or high compression ratios) and then validating the media inside of the zip file using the `dive_dataset/validate_files` endpoint.
- After extracting in the worker the data is pushed to the containing folder.  The zip file has the `ZipFileExtractedMarker` added to it and the `dive_rpc/postprocess` endpoint is called once again to perform any transcoding.
- Second `dive_rpc/postprocess` will also move any zip files that have been processed into a `source` folder to keep the root of the dataset directory clean.
- integration testing was implemented for zip files independently of the other upload/postprocess tests.  New zip package was added to the testing folder for all of these files.  These include tests of malformed zip files and nested zip files, as well as files which contain a high compression ratio (potential zipBombs)

BryonLewis commented:
@subdavis, @marySalvi, @waxlamp  - Soliciting opinions on two things related to uploading Zip files.

- Should uploaded zip files allow for multiple projects to be extracted.  I.E. if you upload a zip file which contains multiple folder and each one of them is a valid dataset (verified using the validation endpoint) should it create and upload each one?
   - I see the argument for having this ability because it may simplify bulk importing.  I also see the counter argument for the complication that it may cause.
- Should a backup of the source zip file be placed in a "SourceZip" folder or should it just stay in the root directory?

BryonLewis commented:
> @subdavis, @marySalvi, @waxlamp - Soliciting opinions on two things related to uploading Zip files.
> 
> * Should uploaded zip files allow for multiple projects to be extracted.  I.E. if you upload a zip file which contains multiple folder and each one of them is a valid dataset (verified using the validation endpoint) should it create and upload each one?
>   
>   * I see the argument for having this ability because it may simplify bulk importing.  I also see the counter argument for the complication that it may cause.
> * Should a backup of the source zip file be placed in a "SourceZip" folder or should it just stay in the root directory?

This was talked about and resulted in the following:
- File and issue for multi-project import using zip, for now stay with a single flat project.
- Keep the zip file in a folder named 'source' that can handle having the data

subdavis commented on server/dive_server/crud_rpc.py:
```diff
@@ -377,6 +378,35 @@ def postprocess(
+            if constants.ZipFileExtractedMarker in item['meta'].keys():
+                source_folder = crud.get_or_create_source_folder(dsFolder, user)
+                Item().move(item, source_folder)
+                continue
+            newjob = tasks.extract_zip.apply_async(
+                queue=_get_queue_name(user),
+                kwargs=dict(
+                    folderId=str(item["folderId"]),
+                    itemId=str(item["_id"]),
+                    girder_job_title=f"Extracting {item['_id']} to folder {str(dsFolder['_id'])}",
+                    girder_client_token=str(token["_id"]),
+                    girder_job_type="private" if job_is_private else "convert",
+                ),
+            )
```
Use "else" instead of continue

subdavis commented on server/dive_server/crud_rpc.py:
```diff
@@ -377,6 +378,35 @@ def postprocess(
+        # only support one zip file found
+        for item in zipItems:
+            # if item is already extracted create/move into 'source' folder
+            if constants.ZipFileExtractedMarker in item['meta'].keys():
+                source_folder = crud.get_or_create_source_folder(dsFolder, user)
+                Item().move(item, source_folder)
+                continue
```
It might make sense to do this as part of the girder worker task: once extract is complete, have girder worker make a move request to the server.  That way you don't need this new metadata at all.

BryonLewis replied:
you weren't thinking of a creating a specific endpoint, right?
More of a create folder, copy item and delete original?:
- `POST /folder`
- `POST /item/id/copy` - (I don't think there is a move)
- `DELETE /item/id/`
or
- `GirderClient.createFolder('sourceFolder')`
- `Girderclient.uploadFile(zipFile)` -again no Move/Copy for Girder Client
- `GirderClient.delete(olditemId)`

subdavis replied:
No new endpoint is needed.  PUT `/item/{id}` will move an item to a different location.

https://viame.kitware.com/api/v1#!/item/item_updateItem

subdavis commented on server/dive_tasks/tasks.py:
```diff
@@ -545,3 +545,59 @@ def convert_images(self: Task, folderId):
+
+            for fileName in listOfFileNames:
+                if fileName.endswith('.zip'):
+                    manager.write("Nested Zip files are invalid!!!!\n")
```
Exclamation points in error messages are a smell, I think.  Anything that might be read by a user that makes them feel "Yelled at" is probably best avoided.

subdavis commented on server/dive_tasks/tasks.py:
```diff
@@ -545,3 +545,59 @@ def convert_images(self: Task, folderId):
+        validation = gc.sendRestRequest(
+            'POST', '/dive_dataset/validate_files', json=listOfFileNames
+        )
+        if 'ok' in validation.keys() and validation['ok'] is True:
```
```suggestion
        if validation.get('ok', False):
```

Python has a lot of features that are just _better_ than JS, and object.get('key', default) is one of them.

subdavis commented on server/dive_tasks/tasks.py:
```diff
@@ -545,3 +545,59 @@ def convert_images(self: Task, folderId):
+            manager.write(f"Media: {validation['media']}\n")
+            dataset_type = validation['type']
+            manager.write(f"Type: {dataset_type}\n")
+            if "ok" in validation and validation["ok"] is False:
```
Why is this condition embedded inside the success condition?  Logic seems out of order.  You can't transition from ERROR to PUSHING OUTPUT

subdavis reviewed (changes requested):
I think there's also a new UI bug:  The "annotations" and "Image Files" (image type only, obvs) lines are now missing from regular video and image sequence.

![Screenshot from 2021-11-29 12-00-23](https://user-images.githubusercontent.com/4214172/143924887-0275ecf9-1608-41c1-a64a-0f289a5327f4.png)
![Screenshot from 2021-11-29 11-59-00](https://user-images.githubusercontent.com/4214172/143924889-8e5049ca-33b4-4216-aab2-b5b421fccf1c.png)

BryonLewis commented:
Addressed comments.
- Left the zip marker in there.  I like the idea of knowing explicitly that this file has been imported, also useful if we add other items into the source folder (like the fact that we don't backup source transcoded images, but do it for source videos)
- added in Job status constants instead of the integer values
- the girder worker task will now do the movement of the zip file to the 'source' folder
- limited the zip file processing to only the first file found in the folder

subdavis commented on client/platform/web-girder/views/Upload.vue:
```diff
@@ -83,11 +100,15 @@ export default defineComponent({
           preUploadErrorMessage.value = null;
           try {
+            if (dstype !== 'zip') {
             // eslint-disable-next-line @typescript-eslint/no-use-before-define
```
Sorry can we do this one too?  I don't know why these got in here, I'm sure there was a reason.

subdavis commented on client/platform/web-girder/views/Upload.vue:
```diff
@@ -70,10 +70,27 @@ export default defineComponent({
+    const openImport = async (dstype: DatasetType | 'zip') => {
       const ret = await openFromDisk(dstype);
       if (!ret.canceled && ret.fileList) {
         // eslint-disable-next-line @typescript-eslint/no-use-before-define
```
And this one?

subdavis commented on server/dive_server/crud_rpc.py:
```diff
@@ -377,6 +378,31 @@ def postprocess(
+            dsFolder,
+            filters={
+                "lowerName": {"$regex": constants.zipRegex},
+                f"meta.{constants.ZipFileExtractedMarker}": {'$exists': False},
```
The item is moved to `source/`, so looking for items where the metadata marker is false has no additinal selection power. Can this be removed?  Or is it just here for defensiveness?

If it's here for defensiveness, I think I'd prefer logic that raises an exception if an impossible condition is encountered.  Maybe remove this from the query, and then `for item in zipItems: if marker is set, raise exception`

Then it will show up as a 500 in the logs and in sentry.

BryonLewis replied:
It's looking for files where the marker doesn't exists, not where it is false but your point is correct.  It's a remnant from before I moved stuff into a source folder.  Maybe I should just remove it.  There is no current use for it, any justification is based on future uses of the /source folder.

subdavis commented on server/dive_tasks/tasks.py:
```diff
@@ -545,3 +545,71 @@ def convert_images(self: Task, folderId):
+
+    with tempfile.TemporaryDirectory() as _working_directory, suppress(utils.CanceledError):
+        _working_directory_path = Path(_working_directory)
+        item: GirderModel = gc.getItem(itemId)
```
If you wanted to be even more careful, you could check the metadata value here in the job and raise an exception if the server has asked you to postprocess an item that was already processed.

subdavis commented on server/dive_tasks/tasks.py:
```diff
@@ -545,3 +545,71 @@ def convert_images(self: Task, folderId):
+                constants.SourceFolderName,
+                reuseExisting=True,
+            )
+            gc.sendRestRequest(
+                "PUT",
+                f"/item/{str(itemId)}?name={item['name']}&folderId={str(created_folder['_id'])}",
+                json=metadata,
+            )
```
I think this PUT and the metadata update should happen just before the directory upload instead of after it.

As is, you could end up in a loop where stuff keeps getting uploaded, but the zip never moves or has its metadata updated, if, for example, an exception is encountered during upload.   In a worst case, this could just run over and over until disk fills up.

Then, if there's a problem, the celery task will report that it failed and it won't be able to run again.

subdavis reviewed (changes requested):
Please reorder operations in the celery task

The rest (defensive logic, eslint issues) are just suggestions you can ignore if you want.

BryonLewis commented:
Importing Zip File Exports:

background:  
An exported dataset in a single zip should also be imported properly into the system.
Contains auxiliary folder which can be ignored except for the most recent result.json file may be inside of it.
Can contain multiple video files (one transcoded and one source)

Logic:
- First check subfolders (exports have a named subfolder which contains the exported data)
- check for a root file in the subfolder "meta.json" - this indicates that this is a valid export zip.
- base on type upload the video or imageData to the girder folder
- If the type is 'video', upload the `video["filename"]` file and set the proper metadata - codec: h264, originalFps: {fps}, orignalFPsString: {fpsString}, source_video: false, transcoder:ffmpeg
- if the type is 'video' also upload the source video and apply the proper metadata - codec:{codec}, originalFps: {fps}, orignalFPsString: {fpsString}, source_video: true
- If the type is 'image-sequence' upload all identified images in the meta.json file
- set the folder metadata - annotate: true, fps: {meta.fps}, type: {image-sequence | video}
- upload the meta.json and the csv
- Call postprocess (if everything is tagged correctly it shouldn't do anything to videos for transcoding),  This should take the csv and convert it to JSON as well as import the meta.json file to set styling and confidence filters.

Erroring:
- If anything in the above logic fails it should drop back to flat file importing.
- So if the meta.json is missing from subfolder (without multiple subfolders it will fall back to standard import)
- Text output in the job need to be updated to indicate that there is no meta.json located in the subfolder so it is treating it as a flat import
- Any misalignment between the meta.json and the contents of the zipFolder should result in an error

subdavis commented:
I'm not totally sure we want to duplicate possibly fragile logic here in the worker task.

* We now have to keep track of another place where metadata gets set if logic changes (it sometimes does), which means more information that only exists in our brains.
* We have new code to worry about breaking if the content of the export zip changes, which it has several times.

I think it'd be better to try and make this function as dumb as possible, leaving the decision making to postprocess where all the hard decisions about what to do with a folder full of unmarked files is already handled.

Once you know you're dealing with an export (meta.json is present) you can just push everything to the server and make the server deal with it.  Postprocess could even pick the dataset type based on what it finds so the worker doesn't need to open the meta file at all.

subdavis commented:
The thing I'm mostly guarding against here is having one logical path mimic the private behavior of another, because that puts a lot of burden on us to remember that we did that.

BryonLewis commented:
This isn't complete yet I still need to double check a bunch of stuff as well as create some integration tests.

Core of what this latest update does:

- dataset export can be used for import now with properly setting metadata based on the meta.json
- multi flat list image/video import
- You can import multiple flat files in a zip file and have them extract to their own folder and become their own datasets.

TODO:
- [ ] Remove auxiliary root folder on multi-imports
- [ ] Clean up flat vs dataset determination when importing datasets
- [ ] Create integration tests for importing multi-flat, datasets, multi-datasets, mixed flat/dataset
- [ ] Test FPS markers as well as more meta.json markers properly.

subdavis commented on server/dive_tasks/tasks.py:
```diff
@@ -545,3 +545,104 @@ def convert_images(self: Task, folderId):
+                manager.write(f"Extracting: {fileName}\n")
+                zipObj.extract(fileName, f'{_working_directory}')
+        # remove the zip file
+        os.remove(file_name)
```
This seems redundant.  The context manager will remove the whole working directory on exit.

BryonLewis replied:
I want to make sure it doesn't re-up the Zip file into the root folder.  If it does you would be an infinite loop of zip extraction.  Other option is to be more explicit about only allowing media file types and .json/.csv types for upload/

subdavis replied:
I see, thanks.

subdavis commented on server/dive_tasks/tasks.py:
```diff
@@ -545,3 +545,104 @@ def convert_images(self: Task, folderId):
+                zipObj.extract(fileName, f'{_working_directory}')
+        # remove the zip file
+        os.remove(file_name)
+        # validation of files in folder using dive/data
+        created_folder = gc.createFolder(
+            folderId,
+            constants.SourceFolderName,
+            reuseExisting=True,
+        )
+        gc.sendRestRequest(
+            "PUT",
+            f"/item/{str(item['_id'])}?folderId={str(created_folder['_id'])}",
+        )
```
Does this comment go with these lines?  Seems like this is moving the zip to source, but the comment says otherwise.

BryonLewis replied:
yeah, I must have moved stuff around and forgot about the comment.

subdavis commented on server/dive_tasks/tasks.py:
```diff
@@ -545,3 +545,104 @@ def convert_images(self: Task, folderId):
+            f"/item/{str(item['_id'])}?folderId={str(created_folder['_id'])}",
+        )
+
+        if len(top_level_folders) > 0:
```
I didn't know you could `len()` a dict, but it works!

BryonLewis replied:
yeah I used to do `len(dict.keys())` until I saw that you could just `len(dict)` and it should get the number of keys.

subdavis commented on server/dive_tasks/tasks.py:
```diff
@@ -545,3 +545,104 @@ def convert_images(self: Task, folderId):
+            utils.upload_zipped_flat_media_files(gc, manager, folderId, Path(_working_directory))
+
+        # remove auxilary folder and metadata if there are sub folders
+        if len(top_level_folders) > 1:
```
What's the point of deleting its auxiliary folder?  Seems fine to me to just remove the metadata and move on.

Also, what's the point of removing anything except for DatasetMarker?  That's the only one that matters.

BryonLewis replied:
It felt weird if you were uploading multiple sub-datasets to have a root folder which contained an Aux folder.  I also didn't want to have a folders lingering around with `type: zip` and `fps` if we did some batch search for a type.  I wanted to behave more like a standard folder with sub datasets that just happens to have a Source folder with the zip in it.

subdavis commented on server/dive_tasks/utils.py:
```diff
@@ -157,3 +159,139 @@ def download_source_media(girder_client: GirderClient, datasetId: str, dest: Pat
+        manager.updateStatus(JobStatus.ERROR)
+
+
+def upload_exported_zipped_dataset(
```
FWIW, this function appears to be quite fragile.  If we find that it breaks often, we may need to revisit this approach, but I think it's fine for now.

subdavis reviewed (changes requested):
There are a few more minor nits here, and one more open ended concern.

I tested with a zip file that extracts to a sub folder (IMO, most correct zip folders should not tarbomb, so I try to make zip files that contain a single sub-folder like this).

It didn't work, and I'm not sure if it is supposed to from looking at the code, but I sorta think it should?

```
Fetching input from 61c0a108ac730287a6f624a8 to /tmp/tmp1cc2rnha/sub_cf2e010daaf375e60dc7728ee74a683f#201106161330_8.zip...
Extracting: sub_cf2e010daaf375e60dc7728ee74a683f#201106161330_8/tmp5p6l0nzv.mp4
Extracting: sub_cf2e010daaf375e60dc7728ee74a683f#201106161330_8/result_04-04-2021_01:55:13.json
Extracting: sub_cf2e010daaf375e60dc7728ee74a683f#201106161330_8/output_tracks.csv
Message: No supported media-type files found
Invalid state transition to '3', Current state is '4'.
```

Also, that invalid transition exists because you shouldn't use `JobStatus.ERROR` directly in a celery task.  You should just raise an exception and let girder worker catch it and do the transition for you.

See examples in other tasks where I `raise Exception`.  You could also do RuntimeError -- I haven't been that consistent.

BryonLewis commented on server/tests/integration/test_zip_uploads.py:
```diff
@@ -0,0 +1,74 @@
+
+@pytest.mark.integration
+@pytest.mark.parametrize("user", zipUser.values())
+@pytest.mark.run(order=4)
```
It seems it would be better to match up user creation, data reset and upload with the other ordering for other tests.

subdavis commented:
I think #1099 is needed for this to work as expected.
